text/XXX-unwind-FFI.md

```diff
@@ -0,0 +1,286 @@
+- Feature Name: `unwind-through-FFI`
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: ???
+- Rust Issue: ???
+
+# Summary
+[summary]: #summary
+
+Provide a well-defined mechanism for unwinding through FFI boundaries.
+
+* Stabilize the function annotation `#[unwind(allowed)]`,
+  which explicitly permits `extern` functions
+  to unwind (`panic`) without aborting.
+* If this annotation is used anywhere in the dependency tree, 
+  generation of the final (binary) product will fail
+  unless the panic strategy is `unwind`
+  and a non-default panic runtime is specified.
+* Stabilize the `#![panic_runtime]` annotation (from
+  [RFC #1513](less-unwinding)).
+* Provide an `unwind` runtime in the standard library
+  that guarantees compatibility with the native exception mechanism
+  provided by the compiler backend.
+* Provide a Cargo option under `[profile.foo]` to specify a `panic` runtime.
+
+[less-unwinding]: https://github.com/rust-lang/rfcs/blob/master/text/1513-less-unwinding.md
+
+# Motivation
+[motivation]: #motivation
+
+This will enable resolving
+[rust-lang/rust#58794](rust-ffi-issue)
+without breaking existing code.
+
+Currently, unwinding through an FFI boundary is always undefined behavior.
+We would like to make the behavior safe
+by aborting when the stack is unwound to an FFI boundary.
+However, there are existing Rust crates
+(notably, wrappers around the `libpng` and `libjpeg` C libraries)
+that rely on the current implementation's compatibility
+with the native exception handling mechanism in
+GCC, LLVM, and MSVC.
+
+This RFC will give crate authors the ability
+to initiate stack-unwinding
+that can propagate through other languages' stack frames,
+as well as tools for ensuring compatibility with those languages.
+
+[rust-ffi-issue]: https://github.com/rust-lang/rust/issues/58794
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Consider an `extern "C"` function that may `panic`:
+
+```rust
+extern "C" fn may_panic(i: i32) {
+    if i < 0 {
+        panic!("Oops, I should have used u32.");
+    }
+}
+```
+
+In a future (TBD) version of Rust,
+calling this function with an argument less than
+zero will cause the program to be aborted.
+This is the only way the compiler can ensure that the function is safe,
+because there is no way for it to know whether or not the calling code
+supports the same implementation of stack-unwinding used for Rust.
```

Comment:
This function is assumed to never `panic!`, and this function, as well as code calling it, will be optimized accordingly.

A valid optimization on this function is therefore to remove the `if i < 0 { panic!() }` brach, since that code cannot ever be reached without invoking undefined behavior. That is, you end up with `extern "C" fn may_panic(i: i32) { }`, and calling that from C code is actually "fine", because no unwinding can happen. 

This section has to somehow be reworked to account for this.

Reply:
Good point. Would adding `opt-level=0` be sufficient?

Reply:
No, undefined behavior is optimization level independent. I think the content in the section is just incorrect and needs to be corrected. As in, the undefined behavior here has nothing to do with FFI. The following snippet has undefined behavior, and no FFI was involved:

```rust
extern "Rust" fn foo() { panic!() } 
fn bar () { foo() }
```

Reply:
@gnzlbg [The Rust reference says](https://doc.rust-lang.org/reference/items/functions.html#extern-functions): "unwinding past the end of an extern function will cause the process to abort", which seems well-defined to me, not UB.

Reply:
You are right, in a future version of Rust that will happen. Right now it is UB. It is unclear to me when the abort happens. I'll ask in the tracking issue.

Reply:
@gnzlbg I'm not sure what point you're trying to make. You mention "a valid optimization" that causes the undefined behavior not to interact with the caller's runtime, but of course `opt-level=0` wouldn't permit such an optimization.

As for the Rust-only, no-FFI version, that's only undefined behavior from the perspective of Rust users because Rust itself doesn't _yet_ define the behavior, and it's only undefined behavior from the perspective of `rustc` itself because of the `nounwind` attribute. The point I am trying to make is that even _without_ the `nounwind` attribute, for an FFI function, there would still be the potential for undefined behavior, because "there is no way for [the compiler] to know whether or not the calling code supports the same implementation of stack-unwinding used for Rust."

Reply:
> @gnzlbg I'm not sure what point you're trying to make. You mention "a valid optimization" that causes the undefined behavior not to interact with the caller's runtime, but of course opt-level=0 wouldn't permit such an optimization.

We do not guarantee that we don't optimize with `-C opt-level=0`, at least not in the language reference. Optimization levels are not part of the language, and undefined behavior is undefined, independently of which optimization level you are using. Otherwise, you could claim for any code that works at `opt-level=0` that it has defined semantics, but this is not the case (one does not follow from the other). 

> The point I am trying to make is that even without the nounwind attribute, for an FFI function, there would still be the potential for undefined behavior, because "there is no way for [the compiler] to know whether or not the calling code supports the same implementation of stack-unwinding used for Rust."

I think the proof obligation here is in the code declaring the function, not the code defining it. Some code _defines_  an `extern "C"` function like this, which is safe:

```rust
extern "C" fn foo() { ... }  // this is safe
```

Calling `foo` from that module is safe. 

However, some other Rust code _declares_ the function to be able to use it:

```rust
extern "C" { fn foo(); }
```

Calling `foo` from this code is `unsafe` because there is no way for the compiler to check whether this declaration matches the definition. 

If you consider the panic-runtimes as part of the function ABI, then in the definition the compiler can make sure that everything is correct, so I don't see why that would need to be unsafe. 

In the declaration, the compiler cannot check anything anyways, so that needs to be unsafe.

Reply:
> Right now it is UB. It is unclear to me when the abort happens. I'll ask in the tracking issue.

The implementation of aborting-at-FFI boundary makes it so that unwinding aborts at the FFI boundary after all Rust frames above the FFI boundary are unwound and all scheduled cleanups are executed.